Added doc-versions page and header link

chupman opened the pull request:
Signed-off-by: Chris Hupman <chupman@us.ibm.com>

This addresses the static content #22 
Preview on my branch https://chupman.github.io/docs.janusgraph.org/latest/index.html

I checked out and ran `./docs/build-and-copy-docs.sh` on the v0.1.0,v0.1.1, and v0.2.0 tags after updating the Other Versions link, adding the doc-versions page and adding it to the appendix. Going forward when versions are added all that will need to be updated is the doc-versions.html file in each release folder.

Since it changed the header it literally touched every static html page for official releases, but it will make user navigation and future maintenance easier. 

Example Diff from the 0.2.0 branch below
```
Chriss-MacBook-Pro-2:janusgraph chupman$ git diff
diff --git a/docs/appendices.txt b/docs/appendices.txt
index 7c1f02198..e4e2dc3d5 100644
--- a/docs/appendices.txt
+++ b/docs/appendices.txt
@@ -16,3 +16,5 @@ include::versions.txt[]
 include::changelog.txt[]

 include::upgrade.txt[]
+
+include::doc-versions.txt[]
diff --git a/docs/xsl/common.xsl b/docs/xsl/common.xsl
index 7ce6766e4..ba0705926 100644
--- a/docs/xsl/common.xsl
+++ b/docs/xsl/common.xsl
@@ -88,7 +88,7 @@
         <ul class="header-list">
           <li class="header-item"><a href="http://janusgraph.org"><img src="images/janusgraph-logo.png" alt="JanusGraph" class="normal_logo" /></a></li>
           <li class="header-item-right"><a href="https://github.com/JanusGraph/janusgraph/releases">Download JanusGraph</a></li>
-          <li class="header-item-right"><a href="https://github.com/JanusGraph/janusgraph/releases">Other Versions</a></li>
+          <li class="header-item-right"><a href="doc-versions.html">Other Doc Versions</a></li>
           <li class="header-item-right"><a href="index.html">Documentation ($MAVEN{project.version})</a></li>
         </ul>
       </header>
```

chupman commented:
all the links other than the appendix link are now hardcoded to https://docs.janusgraph.org/. 

Also the header links point to latest/doc-versions.html instead of a relative version per @mbrukman 's comments in JanusGraph/janusgraph#1046. 
Preview is live at [https://chupman.github.io/docs.janusgraph.org/latest/index.html](https://chupman.github.io/docs.janusgraph.org/latest/index.html)

pluradj commented:
@mbrukman @JanusGraph/committers Could somebody help review these documentation updates? This PR and https://github.com/JanusGraph/janusgraph/pull/1046 have been sitting around for quite a while.

Whether or not we want to go in this direction for the docs, let's get some review and approve/reject so we can move on. It's fine to defer this to a later release as well. No need for this to hold up any releases.

porunov commented on 0.2.0/version-compat.html:
```diff
@@ -29,7 +29,7 @@
 and security exposures with running software that is no longer supported or
 updated. Please check with the software providers to understand their supported
 versions. Users are strongly encouraged to use the latest versions of the
-software.</p><div class="table"><a name="d0e12076"></a><p class="title"><b>Table&nbsp;B.1.&nbsp;Version Compatibility Matrix</b></p><div class="table-contents"><table summary="Version Compatibility Matrix" cellpadding="4px" style="border-collapse: collapse;border-top: 1px solid #1b8669; border-bottom: 1px solid #1b8669; border-left: 1px solid #1b8669; border-right: 1px solid #1b8669; "><colgroup> ...
```
I am kind of confused by the names `d0e12076` and `d0e12079` but nevertheless, why the name has changed from `d0e12076` to `d0e12079` and what does it mean?

chupman replied:
These files were generated by checking out the various release tags and running `docs/build-and-copy-docs.sh`. Since the header was changed and the output is static html pretty much every file was affected.

The changes actually made before the docs were generated are very similar to JanusGraph/janusgraph#1046 except I didn't add the drop-down or the css related to the dropdown. If it helps to clear things up I didn't specifically add the carriage returns or rename d0e12076 to d0e12079. In a sense asciidoc did.

porunov replied:
I just don't know if it is right. Maybe yes, but I think someone has to check it also. I just don't understand why we have carriage return characters on empty lines. Do we need to return carriage if we already did it the line before? Also, it is kind of strange that some files with carriage return symbols and some without.
But really, maybe it doesn't matter. The thing is that carriage return adds a little bit of space, but it's not much. 
Nevertheless, if there would be a concern with files size, then we should just minimize them. As there is no any problem with the load on `docs.janusgraph.org` then we can leave everything as is. 
Nevertheless, whether you keep return carriage characters or remove them as for me both ways are good.

chupman replied:
The svg seems to load without any problems https://chupman.github.io/docs.janusgraph.org/latest/bigtable.html

porunov replied:
There will be no problem to load svg and css files with and without carriage return characters. They both have a special structure where we don't need neither new lines nor carriage return characters. They will both work even in one line. As I said, those invisible characters will take some space (not much) that's all. Those carriage return characters are not needed for Unix systems. They "may" be needed for some Windows systems to read those css and svg files normally. Again, those characters won't have any stability impact on any systems. So, the only question is "Do we need those characters or not?". I assume, if those characters are not added in all css and svg files, then we don't need those characters but I am not sure.
To remove those characters, you can just copy and paste these files from the previous commit and make `git commit --amend`:
```
0.1.0/images/Cloud-Bigtable.svg
0.1.1/images/Cloud-Bigtable.svg
0.2.0/images/Cloud-Bigtable.svg
0.1.0/avada-stub/style.css
0.1.1/avada-stub/style.css
```
But you can leave those characters if you wish. They just add more than 15000 changes in this commit.

chupman replied:
I didn't entirely catch what that particular detail before and was thinking you meant there was an extra carriage return in each of those 6 files. Thank you for spelling it out for me. I now see that 0.1.0/avada-stub/style.css is causing 15000 changes with just carriage returns. I will definitely do an amend later today.

chupman commented:
There was a fix I had make manually specifically for 0.2.0. There were some bugs with a file and it's references getting named incorrectly that was fixed after the release tag. I think there was also another commit to remove a duplicate favicon file that I manually applied. I had some trouble trying to figure out exactly which commit the 0.2.0 docs were actually generated from. 

It might be worthwhile to tag the commit that has all the fixes something like 0.2.0-docs. If we did I would be more than happy to regenerate the 0.2.0 docs off of it.

sjudeng commented:
I tried to validate with v0.1.1 as well. I can checkout that branch and build the docs and they line up with what's in docs.janusgraph.org:master as expected.

```
$ git checkout v0.1.1
$ ./docs/build-and-copy-docs.sh
...
$ (cd ../docs.janusgraph.org; git checkout master)
Already on 'master'
Your branch is up to date with 'origin/master'.
$ diff -uNr ../docs.janusgraph.org/0.1.1 docs.janusgraph.org/0.1.1 | wc -l
       0
```

But when I apply the patch and rebuild and compare to your fork there are still changes. I'm probably missing something, I haven't reviewed doc updates like this before so I don't know the usual process others follow.

```
$ git diff
diff --git a/docs/appendices.txt b/docs/appendices.txt
index 96dc18bba..ca73e305a 100644
--- a/docs/appendices.txt
+++ b/docs/appendices.txt
@@ -16,3 +16,5 @@ include::versions.txt[]
 include::changelog.txt[]
 
 include::upgrade.txt[]
+
+include::doc-versions.txt[]
diff --git a/docs/xsl/common.xsl b/docs/xsl/common.xsl
index a50107417..0f9f3926a 100644
--- a/docs/xsl/common.xsl
+++ b/docs/xsl/common.xsl
@@ -110,7 +110,7 @@
             <nav id="nav" class="nav-holder">
               <ul class="navigation menu fusion-navbar-nav">
                 <li id="menu-item-4711" class="menu-item menu-item-type-custom menu-item-object-custom menu-item-4711"><a href="index.html">Documentation ($MAVEN{project.version})</a></li>
-                <li id="menu-item-4712" class="menu-item menu-item-type-custom menu-item-object-custom menu-item-4712"><a href="https://github.com/JanusGraph/janusgraph/releases">Other Versions</a></li>
+                <li id="menu-item-4712" class="menu-item menu-item-type-custom menu-item-object-custom menu-item-4712"><a href="doc-versions.html">Other Doc Versions</a></li>
                 <li id="menu-item-4710" class="menu-item menu-item-type-custom menu-item-object-custom menu-item-4710"><a href="https://github.com/JanusGraph/janusgraph/releases">Download JanusGraph</a></li>
               </ul>
             </nav>
$ ./docs/build-and-copy-docs.sh
...
$ (cd ../docs.janusgraph.org; git checkout chupman_master)
Switched to branch 'chupman_master'
Your branch is up to date with 'chupman/master'.
$ diff -uNrb ../docs.janusgraph.org/0.1.1 docs.janusgraph.org/0.1.1 | wc -l
     691
```

chupman commented:
@sjudeng I pushed the branches with the changes up on my fork and linked them below. Also I added in some of the changes from JanusGraph/janusgraph#1046 for consistency, but those changes were in docs/doc-versions.txt so it should not have affected the diff.
* [0.1.0](https://github.com/chupman/janusgraph/commit/268a98770c94bdc78e0b2d2dd9ddf08ad2f91751)
* [0.1.1](https://github.com/chupman/janusgraph/commit/6dbc4afa18b8fbd67bc2d6c4c7947948f6476ea2)
* [0.2.0](https://github.com/chupman/janusgraph/commit/5ab9cd66821608ee3fb0e826f4378e76f96185b8)

porunov commented:
@chupman could you please add version 0.2.1 here also?

porunov commented:
@chupman there is also a new version 0.3.0 could you please add it to these PR also?

porunov commented:
@chupman Thank you. Could you also resolve the file `0.2.0/images/Cloud-Bigtable.svg`? It has unneeded  carriage return symbols.

chupman commented:
@porunov thanks for catching the `Cloud-Bigtable.svg` I missed. It's fixed now. I definitely plan to avoid submitting PRs with almost 300 files in the future if at all possible.

porunov commented:
@sjudeng could you please check if your comment was resolved?
There was a big cleanup in this PR.
Seems that this PR is clear now and can be merged if there is no other concerns.

sjudeng commented:
I tried again to reproduce the updates using your above [rel-v0.2.0](https://github.com/chupman/janusgraph/commit/5ab9cd66821608ee3fb0e826f4378e76f96185b8) branch but there are still differences. Is there a better way to cleanly reproduce/validate?

```bash
git fetch chupman
git checkout rel-v0.2.0
./docs/build-and-copy-docs.sh
(cd ../docs.janusgraph.org; git fetch chupman; git reset --hard chupman/master)
diff -uNrb ../docs.janusgraph.org/0.2.0 docs.janusgraph.org/0.2.0 | wc -l
# expect zero differences but get 1053
```

porunov commented:
@sjudeng You are right. I also see those differences.
I've checked the file `0.2.0/advanced-schema.html` and see that in this PR there was added dropdown list of all versions after `Other Doc Versions`:
```
<div class="dropdown-content"><a href="https://docs.janusgraph.org/latest/index.html">Latest</a><a href="https://docs.janusgraph.org/0.3.0/index.html">Version 0.3.0</a><a href="https://docs.janusgraph.org/0.2.1/index.html">Version 0.2.1</a><a href="https://docs.janusgraph.org/0.2.0/index.html">Version 0.2.0</a><a href="https://docs.janusgraph.org/0.1.1/index.html">Version 0.1.1</a><a href="https://docs.janusgraph.org/0.1.0/index.html">Version 0.1.0</a></div>
```
Also, `li` has an additional class `dropdown`:
```
<li class="header-item-right dropdown">
```
But if we try to build those docs from `https://github.com/chupman/janusgraph/tree/rel-v0.2.0` with `./docs/build-and-copy-docs.sh` docs are generated without dropdown list with all versions. I'm confused right now. Don't know how to validate it properly.

chupman commented:
@sjudeng Sorry I hadn't pushed the most recent changes with the dropdown added up to github for rel-v0.1.0,rel-v0.2.0, etc. They are now on my fork for comparison.

In addition there were fixes made to the docs after the tagged release for 0.2.0. I had to manually resolve the issues for 0.2.0 specifically. I suggested earlier in the conversation that we add a tag for the 0.2.0 docs with the fixes that were added later.

Issues I remember manually fixing
field-mappings.html was turned into _field_mappings.html so I had to update filename and all the references. 
favicon.ico was removed.
files without changes that had added carriage returns (thanks @porunov) I did a `git checkout --` to retain the original copies 
```
0.2.0/images/Cloud-Bigtable.svg
0.1.0/avada-stub/style.css
...
```

sjudeng commented:
I still wasn't able to reproduce the 0.2.0 docs in this branch based on the updated rel-v0.2.0 branch. I can take another look if you can provide instructions how to reproduce. But now that you're a committer I think you're approval is implicit and with @pluradj's approval I think you can merge unless there are any objections from others.

chupman commented:
@sjudeng I am a bit curious about the mismatch. When I generate the docs the only two specific things I do are 
1. move the `latest` symbolic link in docs.janusgraph.org to point to the version I'm generating 
2. I pulled `janusgraph` and `docs.janusgraph.org` into `$HOME/github/janusgraph/` and I use that location to generate docs since ` docs/build-and-copy-docs.sh` expects it.

I just generated the docs against this commit and here's a list of the differences that I had to manually resolve.
```
Chriss-MacBook-Pro-2:docs.janusgraph.org chupman$ git status
On branch verion-appendix
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   0.2.0/direct-index-query.html
	modified:   0.2.0/doc-versions.html
	deleted:    0.2.0/field-mapping.html
	modified:   0.2.0/images/Cloud-Bigtable.svg
	modified:   0.2.0/index-backends.html
	modified:   0.2.0/index-parameters.html
	modified:   0.2.0/index.html
	modified:   0.2.0/solr.html
	modified:   latest

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	0.2.0/_field_mapping.html
	0.2.0/favicon.ico

Chriss-MacBook-Pro-2:docs.janusgraph.org chupman$ git diff --stat
 0.2.0/direct-index-query.html   |   6 ++---
 0.2.0/doc-versions.html         |   2 +-
 0.2.0/field-mapping.html        |  44 --------------------------------
 0.2.0/images/Cloud-Bigtable.svg | 156 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------------------
 0.2.0/index-backends.html       |   4 +--
 0.2.0/index-parameters.html     |   6 ++---
 0.2.0/index.html                |   4 +--
 0.2.0/solr.html                 |   4 +--
 latest                          |   2 +-
```
